kore/src/Kore/Step/Function/Evaluator.hs

```diff
@@ -79,50 +93,76 @@ evaluateApplication
+        unevaluatedSimplifier
+          | Just hook <- getHook (Attribute.hook symbolAttributes)
+            -- TODO (thomas.tuegel): (not . null) is a nasty hack to avoid
+            -- making the function evaluator a replaceable component.
+          , (not . null) axiomIdToEvaluator
+          =
+            (error . show . Pretty.vsep)
+                [ "Attempted to evaluate missing hook:" <+> Pretty.pretty hook
+                , "for symbol:" <+> unparse symbol
+                ]
+          | otherwise = return unevaluated
 
-        maybeEvaluatedPattSimplifier =
+        maybeEvaluatedSimplifier =
             maybeEvaluatePattern
                 substitutionSimplifier
                 simplifier
                 axiomIdToEvaluator
                 childrenPredicate
                 termLike
-                unchanged
+                unevaluated
                 configurationPredicate
-        unchangedPatt =
-            Conditional
-                { term         = termLike
-                , predicate    = predicate
-                , substitution = substitution
-                }
-          where
-            Conditional { term = (), predicate, substitution } =
-                childrenPredicate
-        unchanged = OrPattern.fromPattern unchangedPatt
 
-        getSymbolHook = getHook . Attribute.hook . symbolAttributes
-        getAppHookString = Text.unpack <$> getSymbolHook appHead
+    results <- maybeEvaluatedSimplifier & maybe unevaluatedSimplifier Trans.lift
+    Foldable.for_ canMemoize (recordOrPattern results)
+    return results
+  where
+    finishT :: ExceptT r simplifier r -> simplifier r
+    finishT = exceptT return return
+
+    Application { applicationSymbolOrAlias = symbol } = application
+    Symbol { symbolAttributes } = symbol
+
+    termLike = synthesize (ApplySymbolF application)
+    unevaluated =
+        OrPattern.fromPattern $ Pattern.withCondition termLike childrenPredicate
+
+    canMemoize
+      | Symbol.isMemo symbol
+      , isTop childrenPredicate
+      , isTop configurationPredicate
+      = traverse asConcrete application
+      | otherwise
+      = Nothing
+
+    recallOrPattern key = do
+        Memo.Self { recall } <- askMemo
+        maybeTermLike <- recall key
+        let maybeOrPattern =
+                OrPattern.fromTermLike . fromConcrete <$> maybeTermLike
+        Foldable.for_ maybeOrPattern throwE
 
-    case maybeEvaluatedPattSimplifier of
-        Nothing
-          | Just hook <- getAppHookString
-          , not(null axiomIdToEvaluator) ->
-            error
-                (   "Attempting to evaluate unimplemented hooked operation "
-                ++  hook ++ ".\nSymbol: " ++ getIdForError symbolId
-                )
-          | otherwise ->
-            return unchanged
-        Just evaluatedPattSimplifier -> evaluatedPattSimplifier
+    recordOrPattern orPattern key
+      | [result] <- OrPattern.toPatterns orPattern
+      , isTop (Pattern.predicate result)
```

Comment:
1. Why aren't you also testing the substitution? If it's not needed, it would be nice to assert that.
2. Why do you need the predicate to be top? Are you expecting concrete terms to have no attached predicate, so you're only handling that case?